File: urisys/src/urisys/http_server.py
from __future__ import annotations

import json
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlparse

from .controllers.uri_controller import UriController
from .managers.event_manager import EventManager


def _read_json(handler: BaseHTTPRequestHandler) -> dict:
    length = int(handler.headers.get("Content-Length") or 0)
    if not length:
        return {}
    return json.loads(handler.rfile.read(length).decode("utf-8"))


def _send(handler: BaseHTTPRequestHandler, status: int, data: dict) -> None:
    raw = json.dumps(data, ensure_ascii=False, indent=2).encode("utf-8")
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json; charset=utf-8")
    handler.send_header("Access-Control-Allow-Origin", "*")
    handler.send_header("Access-Control-Allow-Methods", "GET,POST,OPTIONS")
    handler.send_header("Access-Control-Allow-Headers", "Content-Type,Authorization")
    handler.send_header("Content-Length", str(len(raw)))
    handler.end_headers()
    handler.wfile.write(raw)


def create_server(host: str, port: int, *, packs="all", markpacts=None, events_path="output/urisys-events.jsonl") -> ThreadingHTTPServer:
    controller = UriController(packs=packs, markpacts=markpacts, events_path=events_path)

    class Handler(BaseHTTPRequestHandler):
        def do_OPTIONS(self):
            _send(self, 204, {})

        def do_GET(self):
            path = urlparse(self.path).path
            if path == "/health":
                _send(self, 200, {"ok": True, "service": "urisys"})
                return
            if path == "/uri/routes":
                _send(self, 200, {"ok": True, "routes": controller.routes()})
                return
            if path == "/uri/events":
                _send(self, 200, {"ok": True, "events": EventManager(events_path).list_events()})
                return
            _send(self, 404, {"ok": False, "error": "not_found", "path": path})

        def do_POST(self):
            path = urlparse(self.path).path
            try:
                body = _read_json(self)
                if path == "/uri/call":
                    context = body.get("context") or {}
                    result = controller.call(
                        body["uri"],
                        body.get("payload") or {},
                        approved=bool(context.get("approved")),
                        dry_run=bool(context.get("dry_run")),
                        allow_real=bool(context.get("allow_real")),
                        environment=str(context.get("environment", "mock")),
                        context=context,
                    )
                    _send(self, 200 if result.get("ok") else 400, result)
                    return
                if path == "/uri/explain":
                    _send(self, 200, {"ok": True, "explain": controller.explain(body["uri"])})
                    return
                _send(self, 404, {"ok": False, "error": "not_found", "path": path})
            except Exception as exc:
                _send(self, 500, {"ok": False, "error": str(exc)})

        def log_message(self, format, *args):
            return

    return ThreadingHTTPServer((host, port), Handler)

File: uricore-js/src/server/http-server.js
import http from 'node:http';
import { readFile, stat } from 'node:fs/promises';
import { createReadStream } from 'node:fs';
import { extname, join, normalize, resolve } from 'node:path';

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.ico': 'image/x-icon',
};

async function readJson(req) {
  const chunks = [];
  for await (const chunk of req) chunks.push(chunk);
  const raw = Buffer.concat(chunks).toString('utf8') || '{}';
  return JSON.parse(raw);
}

function sendJson(res, status, data, cors) {
  if (cors) setCors(res);
  res.writeHead(status, { 'Content-Type': 'application/json; charset=utf-8' });
  res.end(JSON.stringify(data, null, 2));
}

function setCors(res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET,POST,OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type,Authorization');
}

export function createHttpUriServer({ runtime, staticDir = null, cors = true } = {}) {
  if (!runtime) throw new Error('createHttpUriServer requires runtime.');
  const sseClients = new Set();

  runtime.eventStore.subscribe?.((event) => {
    const payload = `event: uricore.event\ndata: ${JSON.stringify(event)}\n\n`;
    for (const res of sseClients) res.write(payload);
  });

  const server = http.createServer(async (req, res) => {
    try {
      if (cors) setCors(res);
      if (req.method === 'OPTIONS') {
        res.writeHead(204);
        res.end();
        return;
      }

      const url = new URL(req.url, `http://${req.headers.host || 'localhost'}`);

      if (req.method === 'GET' && url.pathname === '/health') {
        sendJson(res, 200, { ok: true, service: 'uricore-js' }, cors);
        return;
      }

      if (req.method === 'GET' && url.pathname === '/uri/capabilities') {
        sendJson(res, 200, { ok: true, capabilities: runtime.registry.listCapabilities() }, cors);
        return;
      }

      if (req.method === 'GET' && url.pathname === '/uri/events') {
        const since = Number(url.searchParams.get('since') || 0);
        const limit = Number(url.searchParams.get('limit') || 100);
        sendJson(res, 200, { ok: true, events: runtime.eventStore.list({ since, limit }) }, cors);
        return;
      }

      if (req.method === 'GET' && url.pathname === '/uri/sse') {
        res.writeHead(200, {
          'Content-Type': 'text/event-stream',
          'Cache-Control': 'no-cache',
          Connection: 'keep-alive',
          'Access-Control-Allow-Origin': '*',
        });
        res.write(`event: uricore.ready\ndata: ${JSON.stringify({ ok: true })}\n\n`);
        sseClients.add(res);
        req.on('close', () => sseClients.delete(res));
        return;
      }

      if (req.method === 'POST' && (url.pathname === '/uri/call' || url.pathname === '/uri/query')) {
        const body = await readJson(req);
        const result = await runtime.call(body.uri, body.payload || {}, body.context || {});
        sendJson(res, result.ok ? 200 : 400, result, cors);
        return;
      }

      if (staticDir && req.method === 'GET') {
        const served = await tryServeStatic(staticDir, url.pathname, res);
        if (served) return;
      }

      sendJson(res, 404, { ok: false, error: 'not_found', path: url.pathname }, cors);
    } catch (error) {
      sendJson(res, 500, { ok: false, error: error.message }, cors);
    }
  });

  return server;
}

async function tryServeStatic(staticDir, pathname, res) {
  const root = resolve(staticDir);
  const requested = pathname === '/' ? '/index.html' : pathname;
  const file = normalize(join(root, decodeURIComponent(requested)));
  if (!file.startsWith(root)) return false;
  try {
    const info = await stat(file);
    if (!info.isFile()) return false;
    res.writeHead(200, { 'Content-Type': MIME[extname(file)] || 'application/octet-stream' });
    createReadStream(file).pipe(res);
    return true;
  } catch {
    return false;
  }
}
